drizzle-valibot/src/index.ts

```diff
@@ -38,58 +45,69 @@ import {
 	picklist,
 	type PicklistSchema,
 	record,
+	type RecordIssue,
+	pipe,
+	type SchemaWithPipe,
 	string,
+	type StringIssue,
 	type StringSchema,
 	union,
+	type UnionIssue,
 	uuid,
 } from 'valibot';
 
-const literalSchema = union([string(), number(), boolean(), null_()]);
+// TODO: I think types test would be great for this file to make sure types are inferred correctly
 
-type Json = typeof jsonSchema;
-
-export const jsonSchema = union([literalSchema, array(any()), record(any())]);
+type Json = string | number | boolean | null | { [key: string]: Json } | Json[];
+type LiteralIssue = StringIssue | NumberIssue | BooleanIssue | NullIssue;
+type JsonIssue = LiteralIssue | ArrayIssue | RecordIssue;
+type JsonSchema = BaseSchema<Json, Json, UnionIssue<JsonIssue> | JsonIssue>;
+export const jsonSchema = lazy(() =>
+	union([string(), number(), boolean(), null_(), array(jsonSchema), record(string(), jsonSchema)])
+) as JsonSchema;
 
 type MapInsertColumnToValibot<
 	TColumn extends Column,
-	TType extends BaseSchema<any, any>,
-> = TColumn['_']['notNull'] extends false ? OptionalSchema<NullableSchema<TType>>
-	: TColumn['_']['hasDefault'] extends true ? OptionalSchema<TType>
+	TType extends BaseSchema<unknown, unknown, BaseIssue<unknown>>,
+> = TColumn['_']['notNull'] extends false ? OptionalSchema<NullableSchema<TType, never>, never>
+	: TColumn['_']['hasDefault'] extends true ? OptionalSchema<TType, never>
 	: TType;
 
 type MapSelectColumnToValibot<
 	TColumn extends Column,
-	TType extends BaseSchema<any, any>,
-> = TColumn['_']['notNull'] extends false ? NullableSchema<TType> : TType;
+	TType extends BaseSchema<unknown, unknown, BaseIssue<unknown>>,
+> = TColumn['_']['notNull'] extends false ? NullableSchema<TType, never> : TType;
 
 type MapColumnToValibot<
 	TColumn extends Column,
-	TType extends BaseSchema<any, any>,
+	TType extends BaseSchema<unknown, unknown, BaseIssue<unknown>>,
 	TMode extends 'insert' | 'select',
 > = TMode extends 'insert' ? MapInsertColumnToValibot<TColumn, TType>
 	: MapSelectColumnToValibot<TColumn, TType>;
 
 type MaybeOptional<
 	TColumn extends Column,
-	TType extends BaseSchema<any, any>,
+	TType extends BaseSchema<unknown, unknown, BaseIssue<unknown>>,
 	TMode extends 'insert' | 'select',
 	TNoOptional extends boolean,
 > = TNoOptional extends true ? TType
 	: MapColumnToValibot<TColumn, TType, TMode>;
 
 type GetValibotType<TColumn extends Column> = TColumn['_']['dataType'] extends infer TDataType
 	? TDataType extends 'custom' ? AnySchema
-	: TDataType extends 'json' ? Json
+	: TDataType extends 'json' ? JsonSchema
 	: TColumn extends { enumValues: [string, ...string[]] }
-		? Equal<TColumn['enumValues'], [string, ...string[]]> extends true ? StringSchema
-		: PicklistSchema<TColumn['enumValues']>
+		// TODO: Can we somehow check `.length` to know if it's a string schema with or without a max length check?
```

Comment:
What do you want to get here exacly?
In theory the following should work, but there is a chance that typescript no longer has the exact information of `T["length"]` and it has just become `number`.
```ts
type MaybeLength<T> = T["length"] extends 0 | undefined ? NoLengthType<0> : LengthType<T["length"]>
```

Reply:
I was just trying to be more specific about the returned type. I think the Drizzle core team knows how to proceed and are free to modify the code as I am busy with other things.